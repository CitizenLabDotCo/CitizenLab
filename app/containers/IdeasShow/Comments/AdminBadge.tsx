import React from 'react';
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';
import styled from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';
import { lighten } from 'polished';

const Container = styled.span`
  color: ${colors.clRed};
  font-size: ${fontSizes.xs}px;
  line-height: 16px;
  border-radius: 5px;
  text-transform: uppercase;
  text-align: center;
  font-weight: 600;
  background-color: ${lighten(.45, colors.clRed)};
  border: none;
  padding: 4px 8px;
  height: 24px;
  display: flex;
  align-items: center;
`;

interface Props {
  className?: string;
}

export default (props: Props) => (
  <Container className={props.className}>
    <FormattedMessage {...messages.official} />
  </Container>
);
