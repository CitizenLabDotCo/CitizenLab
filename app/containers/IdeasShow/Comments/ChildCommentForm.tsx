// libraries
import React, { PureComponent, FormEvent } from 'react';
import { Subscription } from 'rxjs';
import { filter, tap } from 'rxjs/operators';
import { trim } from 'lodash-es';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';
import scrollToComponent from 'react-scroll-to-component';

// components
import Button from 'components/UI/Button';
import MentionsTextArea from 'components/UI/MentionsTextArea';
import Observer from '@researchgate/react-intersection-observer';

// tracking
import { trackEventByName } from 'utils/analytics';
import tracks from './tracks';

// i18n
import { InjectedIntlProps } from 'react-intl';
import { injectIntl, FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

// services
import { addCommentToComment } from 'services/comments';
import eventEmitter from 'utils/eventEmitter';

// resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetAuthUser, { GetAuthUserChildProps } from 'resources/GetAuthUser';

// style
import styled from 'styled-components';
import { hideVisually } from 'polished';

// typings
import { ICommentReplyClicked } from './CommentFooter';

const Container = styled.form`
  background: #fff;
  border: solid 1px #fff;
  border-top-color: #ebebeb;
  border-bottom-left-radius: 3px;
  border-bottom-right-radius: 3px;
  transition: all 100ms ease;

  &.focussed {
    background: #fff;
    border-color: #999;
  }
`;

const HiddenLabel = styled.span`
  ${hideVisually()}
`;

const FormInner = styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  padding: 30px 50px;
`;

const TextareaWrapper = styled.div``;

const ButtonWrapper = styled.div`
  display: none;
  justify-content: flex-end;

  &.visible {
    display: flex;
  }
`;

interface InputProps {
  ideaId: string;
  parentId: string;
  className?: string;
}

interface DataProps {
  locale: GetLocaleChildProps;
  authUser: GetAuthUserChildProps;
}

interface Props extends InputProps, DataProps {}

interface State {
  inputValue: string;
  focussed: boolean;
  processing: boolean;
  errorMessage: string | null;
  canSubmit: boolean;
}

class ChildCommentForm extends PureComponent<Props & InjectedIntlProps, State> {
  textareaElement: HTMLTextAreaElement;
  isInViewport: boolean;
  subscriptions: Subscription[] = [];

  constructor(props) {
    super(props);
    this.state = {
      inputValue: '',
      focussed: false,
      processing: false,
      errorMessage: null,
      canSubmit: false,
    };
  }

  componentDidMount() {
    this.subscriptions = [
      eventEmitter.observeEvent<ICommentReplyClicked>('commentReplyButtonClicked').pipe(
        tap(() => {
          this.setState({ inputValue: '', focussed: false });
        }),
        filter(({ eventValue }) => {
          const { commentId, parentCommentId } = eventValue;
          return (commentId === this.props.parentId || parentCommentId === this.props.parentId);
        })
      ).subscribe(({ eventValue }) => {
        const { authorFirstName, authorLastName, authorSlug } = eventValue;
        const inputValue = `@[${authorFirstName} ${authorLastName}](${authorSlug}) `;
        this.setState({ inputValue, focussed: true });

        if (this.textareaElement) {
          // this.textareaElement.scrollIntoView();
          if (this.isInViewport) {
            this.textareaElement.focus();
          } else {
            // this.textareaElement.scrollIntoView({behavior: 'smooth', block: 'end', inline: 'center'});
            scrollToComponent(this.textareaElement, { align: 'top', offset: -400, duration: 400 });

            setTimeout(() => {
              this.textareaElement.focus();
            }, 400);
          }
        }
      })
    ];
  }

  componentWillUnmount() {
    this.subscriptions.forEach(subscription => subscription.unsubscribe());
  }

  handleTextareaOnChange = (inputValue: string) => {
    this.setState(({ focussed }) => ({
      inputValue,
      errorMessage: null,
      canSubmit: !!(focussed && trim(inputValue) !== '')
    }));
  }

  handleTextareaOnFocus = () => {
    trackEventByName(tracks.focusChildCommentEditor, {
      extra: {
        ideaId: this.props.ideaId,
        parentId: this.props.parentId
      }
    });

    this.setState({ focussed: true });
  }

  handleTextareaOnBlur = () => {
    this.setState({ focussed: false });
  }

  handleSubmit = async (event: FormEvent<any>) => {
    event.preventDefault();

    const { locale, authUser, ideaId, parentId } = this.props;
    const { formatMessage } = this.props.intl;
    const { inputValue, canSubmit } = this.state;

    if (!isNilOrError(locale) && !isNilOrError(authUser) && canSubmit) {
      this.setState({
        processing: true,
        canSubmit: false
      });

      trackEventByName(tracks.clickChildCommentPublish, {
        extra: {
          ideaId,
          parentId,
          content: inputValue,
        }
      });

      try {
        await addCommentToComment(ideaId, authUser.id, parentId, { [locale]: inputValue.replace(/\@\[(.*?)\]\((.*?)\)/gi, '@$2') });

        this.setState({
          inputValue: '',
          processing: false
        });
      } catch (error) {
        this.setState({
          errorMessage: formatMessage(messages.addCommentError),
          processing: false,
          canSubmit: true
        });
      }
    }
  }

  setRef = (element: HTMLTextAreaElement) => {
    this.textareaElement = element;
  }

  handleIntersection = (event: IntersectionObserverEntry) => {
    this.isInViewport = event.isIntersecting;
  }

  render() {
    const { ideaId, authUser, className } = this.props;

    if (!isNilOrError(authUser)) {
      const { formatMessage } = this.props.intl;
      const { inputValue, canSubmit, processing, errorMessage, focussed } = this.state;
      const placeholder = formatMessage(messages.childCommentBodyPlaceholder);
      const isButtonVisible = (inputValue && inputValue.length > 0 || focussed);

      return (
        <Container className={`${className} ${focussed ? 'focussed' : ''}`} onSubmit={this.handleSubmit}>
          <Observer onChange={this.handleIntersection}>
            <label>
              <HiddenLabel>
                <FormattedMessage {...messages.replyToComment} />
              </HiddenLabel>
              <FormInner>
                <TextareaWrapper>
                  <MentionsTextArea
                    name="comment"
                    className={`e2e-reply childcommentform-${this.props.parentId}`}
                    placeholder={placeholder}
                    rows={1}
                    value={inputValue}
                    error={errorMessage}
                    ideaId={ideaId}
                    onChange={this.handleTextareaOnChange}
                    onFocus={this.handleTextareaOnFocus}
                    onBlur={this.handleTextareaOnBlur}
                    getTextareaRef={this.setRef}
                    fontWeight="300"
                    padding="10px 0px"
                    borderRadius="none"
                    border="none"
                    boxShadow="none"
                  />
                </TextareaWrapper>
                <ButtonWrapper className={isButtonVisible ? 'visible' : ''}>
                  <Button
                    className="e2e-submit-comment"
                    processing={processing}
                    icon="send"
                    onClick={this.handleSubmit}
                    disabled={!canSubmit}
                  >
                    <FormattedMessage {...messages.publishComment} />
                  </Button>
                </ButtonWrapper>
              </FormInner>
            </label>
          </Observer>
        </Container>
      );
    }

    return null;
  }
}

const ChildCommentFormWithHoCs = injectIntl<Props>(ChildCommentForm);

const Data = adopt<DataProps, InputProps>({
  locale: <GetLocale />,
  authUser: <GetAuthUser />
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <ChildCommentFormWithHoCs {...inputProps} {...dataProps} />}
  </Data>
);
