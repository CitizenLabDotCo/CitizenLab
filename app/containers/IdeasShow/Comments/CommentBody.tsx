// Libraries
import React, { PureComponent, FormEvent } from 'react';
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { get } from 'lodash-es';
import linkifyHtml from 'linkifyjs/html';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';

// Router
import Link from 'utils/cl-router/Link';

// Services
import { updateComment, IUpdatedComment } from 'services/comments';
import eventEmitter from 'utils/eventEmitter';

// Resources
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';
import GetTenantLocales, { GetTenantLocalesChildProps } from 'resources/GetTenantLocales';
import GetComment, { GetCommentChildProps } from 'resources/GetComment';
import GetUser, { GetUserChildProps } from 'resources/GetUser';
import GetMachineTranslation from 'resources/GetMachineTranslation';

// i18n
import { getLocalized } from 'utils/i18n';
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

// Components
import UserName from 'components/UI/UserName';
import MentionsTextArea from 'components/UI/MentionsTextArea';
import Button from 'components/UI/Button';
import Error from 'components/UI/Error';
import QuillEditedContent from 'components/UI/QuillEditedContent';

// Styling
import styled from 'styled-components';
import { darken } from 'polished';
import { colors, media } from 'utils/styleUtils';

// Typings
import { CLErrorsJSON, CLErrors } from 'typings';

const Container = styled.div``;

const CommentWrapper = styled.div`
  white-space: pre-line;

  &.child {
    margin-top: 7px;
  }
`;

const StyledLink = styled(Link)`
  display: inline;
  margin-right: 10px;
  text-decoration: none !important;

  ${media.smallerThanMinTablet`
    display: none;
  `}
`;

const StyledUserName = styled(UserName)`
  color: ${({ theme }) => theme.colorText};
  font-weight: 400;
  text-decoration: none !important;
  cursor: pointer;

  &:hover {
    color: ${({ theme }) => darken(0.15, theme.colorText)};
    text-decoration: underline !important;
  }

  &.canModerate {
    color: ${colors.clRed};

    &:hover {
      color: ${darken(0.15, colors.clRed)};
    }
  }
`;

const CommentText = styled.div`
  display: inline;
`;

const StyledForm = styled.form`
  position: relative;
`;

const ButtonsWrapper = styled.div`
  display: flex;
  justify-content: flex-end;
  margin-top: 1em;

  > * + * {
    margin-left: .5rem;
  }
`;

interface InputProps {
  commentId: string;
  commentType: 'parent' | 'child';
  moderator: boolean;
  editing: boolean;
  last?: boolean;
  onCommentSaved: () => void;
  onCancelEditing: () => void;
  className?: string;
}

interface DataProps {
  locale: GetLocaleChildProps;
  tenantLocales: GetTenantLocalesChildProps;
  comment: GetCommentChildProps;
  author: GetUserChildProps;
}

interface Props extends InputProps, DataProps {}

export interface State {
  commentContent: string;
  editableCommentContent: string;
  translateButtonClicked: boolean;
  processing: boolean;
  apiErrors: CLErrors | null;
}

class CommentBody extends PureComponent<Props, State> {
  subscriptions: Subscription[] = [];

  constructor(props) {
    super(props);
    this.state = {
      commentContent: '',
      editableCommentContent: '',
      translateButtonClicked: false,
      processing: false,
      apiErrors: null
    };
  }

  componentDidMount() {
    this.setCommentContent();
    this.setEditableCommentContent();

    this.subscriptions = [
      eventEmitter.observeEvent<string>('commentTranslateButtonClicked').pipe(
        filter(({ eventValue: commentId }) => commentId === this.props.commentId)
      ).subscribe(() => {
        this.setState(({ translateButtonClicked }) => ({ translateButtonClicked: !translateButtonClicked }));
      })
    ];
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.comment !== this.props.comment) {
      this.setCommentContent();
      this.setEditableCommentContent();
    }
  }

  componentWillUnmount() {
    this.subscriptions.forEach(subscription => subscription.unsubscribe());
  }

  setCommentContent = () => {
    let commentContent = '';
    const { comment, locale, tenantLocales } = this.props;

    if (!isNilOrError(locale) && !isNilOrError(tenantLocales) && !isNilOrError(comment)) {
      commentContent = linkifyHtml(getLocalized(comment.attributes.body_multiloc, locale, tenantLocales).replace(
        /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>([\S\s]*?)<\/span>/gi,
        '<a class="mention" data-link="/profile/$2" href="/profile/$2">$3</a>'
      )) as string;
    }

    this.setState({ commentContent });
  }

  setEditableCommentContent = () => {
    let editableCommentContent = '';
    const { comment, locale, tenantLocales } = this.props;

    if (!isNilOrError(locale) && !isNilOrError(tenantLocales) && !isNilOrError(comment)) {
      editableCommentContent = getLocalized(comment.attributes.body_multiloc, locale, tenantLocales).replace(
        /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>@([\S\s]*?)<\/span>/gi,
        '@[$3]($2)'
      );
    }

    this.setState({ editableCommentContent });
  }

  onEditableCommentContentChange = (editableCommentContent: string) => {
    this.setState({ editableCommentContent });
  }

  onSubmit = async (event: FormEvent<any>) => {
    event.preventDefault();

    const { locale, commentId } = this.props;
    const { editableCommentContent } = this.state;

    if (!isNilOrError(locale)) {
      const updatedComment: IUpdatedComment = {
        body_multiloc: {
          [locale]: editableCommentContent.replace(/\@\[(.*?)\]\((.*?)\)/gi, '@$2')
        }
      };

      this.setState({ processing: true, apiErrors: null });

      try {
        await updateComment(commentId, updatedComment);
        this.props.onCommentSaved();
      } catch (error) {
        if (error && error.json) {
          const apiErrors = (error as CLErrorsJSON).json.errors;
          this.setState({ apiErrors });
        }
      }

      this.setState({ processing: false });
    }
  }

  cancelEditing = (event: FormEvent<HTMLButtonElement>) => {
    event.preventDefault();
    this.setEditableCommentContent();
    this.props.onCancelEditing();
  }

  render() {
    const {
      editing,
      commentType,
      moderator,
      locale,
      author,
      commentId,
      className
    } = this.props;

    const {
      commentContent,
      editableCommentContent,
      translateButtonClicked,
      processing,
      apiErrors
    } = this.state;

    let content: JSX.Element | null = null;

    if (!isNilOrError(locale)) {
      if (!editing) {
        const CommentBodyContent = ({ text }: { text: string }) => (
          <>
            {commentType === 'child' &&
              <StyledLink to={!isNilOrError(author) ? `/profile/${author.attributes.slug}` : ''}>
                <StyledUserName
                  className={moderator ? 'canModerate' : ''}
                  user={!isNilOrError(author) ? author : null}
                />
              </StyledLink>
            }
            <CommentText dangerouslySetInnerHTML={{ __html: text }} />
          </>
        );

        content = (
          <CommentWrapper className={`e2e-comment-body ${commentType}`}>
            <QuillEditedContent fontWeight={300}>
              {translateButtonClicked ? (
                <GetMachineTranslation attributeName="body_multiloc" localeTo={locale} commentId={commentId}>
                  {translation => {
                    let text: string = commentContent;

                    if (!isNilOrError(translation)) {
                      text = translation.attributes.translation.replace(
                        /<span\sclass="cl-mention-user"[\S\s]*?data-user-id="([\S\s]*?)"[\S\s]*?data-user-slug="([\S\s]*?)"[\S\s]*?>([\S\s]*?)<\/span>/gi,
                        '<a class="mention" data-link="/profile/$2" href="/profile/$2">$3</a>'
                      );
                    }

                    return <CommentBodyContent text={text} />;
                  }}
                </GetMachineTranslation>
              ) : (
                <CommentBodyContent text={commentContent} />
              )}
            </QuillEditedContent>
          </CommentWrapper>
        );
      } else {
        content = (
          <StyledForm onSubmit={this.onSubmit}>
            <QuillEditedContent fontWeight={300}>
              <MentionsTextArea
                name="body"
                value={editableCommentContent}
                rows={1}
                onChange={this.onEditableCommentContentChange}
                padding="15px"
                fontWeight="300"
              />
            </QuillEditedContent>
            <ButtonsWrapper>
              {apiErrors && apiErrors.body_multiloc && apiErrors.body_multiloc[locale] &&
                <Error apiErrors={apiErrors.body_multiloc[locale]} />
              }
              <Button
                style="secondary"
                onClick={this.cancelEditing}
              >
                <FormattedMessage {...messages.cancelCommentEdit} />
              </Button>
              <Button
                style="primary"
                processing={processing}
                onClick={this.onSubmit}
              >
                <FormattedMessage {...messages.saveCommentEdit} />
              </Button>
            </ButtonsWrapper>
          </StyledForm>
        );
      }

      return (
        <Container className={className}>
          {content}
        </Container>
      );
    }

    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  locale: <GetLocale />,
  tenantLocales: <GetTenantLocales />,
  comment: ({ commentId, render }) => <GetComment id={commentId}>{render}</GetComment>,
  author: ({ comment, render }) => <GetUser id={get(comment, 'relationships.author.data.id')}>{render}</GetUser>
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <CommentBody {...inputProps} {...dataProps} />}
  </Data>
);
