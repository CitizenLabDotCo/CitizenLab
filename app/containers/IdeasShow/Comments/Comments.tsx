import React, { PureComponent } from 'react';
import { isNilOrError } from 'utils/helperUtils';

// components
import ParentComment from './ParentComment';

// resources
import GetComments, { GetCommentsChildProps } from 'resources/GetComments';

// style
import styled from 'styled-components';

const Container = styled.div`
  padding-bottom: 80px;
`;

interface InputProps {
  ideaId: string;
  className?: string;
}

interface DataProps {
  comments: GetCommentsChildProps;
}

interface Props extends InputProps, DataProps {}

interface State {}

class Comments extends PureComponent<Props, State> {
  render() {
    const { ideaId, comments, className } = this.props;

    if (!isNilOrError(comments) && comments.length > 0) {
      const parentComments = comments.filter((comment) => {
        return comment.relationships.parent.data === null;
      }).sort((commentA, commentB) => {
        return new Date(commentA.attributes.created_at).getTime() - new Date(commentB.attributes.created_at).getTime();
      });

      if (parentComments && parentComments.length > 0) {
        return (
          <Container className={`e2e-comments-container ${className}`}>
            {parentComments.map((parentComment, _index) => (
              <ParentComment
                key={parentComment.id}
                ideaId={ideaId}
                commentId={parentComment.id}
              />
            ))}
          </Container>
        );
      }
    }

    return null;
  }
}

export default (inputProps: InputProps) => (
  <GetComments ideaId={inputProps.ideaId}>
    {comments => <Comments {...inputProps} comments={comments} />}
  </GetComments>
);
