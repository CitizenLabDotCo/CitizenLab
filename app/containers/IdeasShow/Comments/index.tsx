// libraries
import React, { memo, useState, useCallback } from 'react';
import { get } from 'lodash-es';
import { adopt } from 'react-adopt';
import Observer from '@researchgate/react-intersection-observer';

// services
import { canModerate } from 'services/permissions/rules/projectPermissions';

// resources
import GetAuthUser, { GetAuthUserChildProps } from 'resources/GetAuthUser';
import GetIdea, { GetIdeaChildProps } from 'resources/GetIdea';
import GetProject, { GetProjectChildProps } from 'resources/GetProject';
import GetComments, { GetCommentsChildProps } from 'resources/GetComments';

// utils
import { isNilOrError } from 'utils/helperUtils';

// components
import LoadingComments from './LoadingComments';
import ParentCommentForm from './ParentCommentForm';
import Comments from './Comments';
import CommentingDisabled from './CommentingDisabled';
import Warning from 'components/UI/Warning';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

// style
import styled from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';

// typings
import { CommentsSort } from 'services/comments';

const Container = styled.div``;

const StyledWarning = styled(Warning)`
  margin-bottom: 20px;
`;

const LoadMore = styled.div`
  width: 100%;
  height: 0px;
`;

const LoadingMore = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 30px;
`;

const LoadingMoreMessage = styled.div`
  color: ${colors.label};
  font-size: ${fontSizes.medium}px;
  font-weight: 400;
`;

export interface InputProps {
  ideaId: string;
  className?: string;
}

interface DataProps {
  authUser: GetAuthUserChildProps;
  idea: GetIdeaChildProps;
  comments: GetCommentsChildProps;
  project: GetProjectChildProps;
}

interface Props extends InputProps, DataProps {}

const CommentsSection = memo<Props>(({ ideaId, authUser, idea, comments, project, className }) => {
  const [sortOrder, setSortOrder] = useState<CommentsSort>('-new');
  const [posting, setPosting] = useState(false);
  const { commentsList, hasMore, onLoadMore, loadingInital, loadingMore, onChangeSort } = comments;

  const handleSortOrderChange = useCallback(
    (sortOrder: CommentsSort) => {
      onChangeSort(sortOrder);
      setSortOrder(sortOrder);
    }, []
  );

  const handleIntersection = useCallback(
    (event: IntersectionObserverEntry, unobserve: () => void) => {
      if (event.isIntersecting) {
        onLoadMore();
        unobserve();
      }
    }, []
  );

  const handleCommentPosting = useCallback(
    (isPosting: boolean) => {
      setPosting(isPosting);
    }, []
  );

  const isModerator = !isNilOrError(authUser) && canModerate(get(project, 'id'), { data: authUser });
  const commentingEnabled = (!isNilOrError(idea) ? get(idea.relationships.action_descriptor.data.commenting, 'enabled', false) : false);
  const commentingDisabledReason = (!isNilOrError(idea) ? get(idea.relationships.action_descriptor.data.commenting, 'disabled_reason', null) : null);

  return (
    <Container className={className}>
      {(!isNilOrError(idea) && !isNilOrError(commentsList) && !isNilOrError(project)) ? (
        <>
          {/*
            Show warning messages when there are no comments and you're looged in as an admin.
            Otherwise the comment section would be empty (because admins don't see the parent comment box), which might look weird or confusing
          */}
          {isModerator && commentsList && commentsList.length === 0 && !commentingDisabledReason &&
            <StyledWarning>
              <FormattedMessage {...messages.noComments} />
            </StyledWarning>
          }

          <CommentingDisabled
            isLoggedIn={!!authUser}
            commentingEnabled={commentingEnabled}
            commentingDisabledReason={commentingDisabledReason}
            projectId={project.id}
          />

          <Comments
            ideaId={ideaId}
            comments={commentsList}
            sortOrder={sortOrder}
            loading={loadingInital}
            onSortOrderChange={handleSortOrderChange}
          />

          {hasMore && !loadingMore &&
            <Observer onChange={handleIntersection} rootMargin="3000px">
              <LoadMore />
            </Observer>
          }

          {loadingMore && !posting &&
            <LoadingMore>
              <LoadingMoreMessage>
                <FormattedMessage {...messages.loadingMoreComments} />
              </LoadingMoreMessage>
            </LoadingMore>
          }

          <ParentCommentForm
            ideaId={ideaId}
            postingComment={handleCommentPosting}
          />
        </>
      ) : (
        <LoadingComments />
      )}
    </Container>
  );
});

const Data = adopt<DataProps, InputProps>({
  authUser: <GetAuthUser />,
  idea: ({ ideaId, render }) => <GetIdea id={ideaId}>{render}</GetIdea>,
  comments: ({ ideaId, render }) => <GetComments ideaId={ideaId}>{render}</GetComments>,
  project: ({ idea, render }) => <GetProject id={get(idea, 'relationships.project.data.id')}>{render}</GetProject>
});

export default memo<InputProps>((inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <CommentsSection {...inputProps} {...dataProps} />}
  </Data>
));
