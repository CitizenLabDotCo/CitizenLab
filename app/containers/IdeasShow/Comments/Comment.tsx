// libraries
import React, { PureComponent } from 'react';
import { adopt } from 'react-adopt';
import { isNilOrError } from 'utils/helperUtils';
import { get } from 'lodash-es';
import clHistory from 'utils/cl-router/history';

// components
import CommentHeader from './CommentHeader';
import CommentBody from './CommentBody';
import CommentFooter from './CommentFooter';
import Avatar from 'components/Avatar';
import Icon from 'components/UI/Icon';

// services
import { canModerate } from 'services/permissions/rules/projectPermissions';

// resources
import GetComment, { GetCommentChildProps } from 'resources/GetComment';
import GetUser, { GetUserChildProps } from 'resources/GetUser';

// analytics
import { trackEventByName } from 'utils/analytics';
import tracks from './tracks';

// i18n
import { FormattedMessage } from 'utils/cl-intl';
import messages from '../messages';

// style
import styled from 'styled-components';
import { media, colors, fontSizes } from 'utils/styleUtils';

const Container = styled.div`
  &.child {
    background: #fbfbfb;
    border-bottom-left-radius: 3px;
    border-bottom-right-radius: 3px;
  }
`;

const ContainerInner = styled.div`
  display: flex;
  flex-direction: column;
  align-items: stretch;
  padding-top: 28px;
  padding-bottom: 25px;

  &.hasBottomBorder {
    border-bottom: solid 1px #e8e8e8;
  }

  &.lastComment {
    border-bottom: none;
    padding-bottom: 40px;
  }

  &.parent {
    padding-left: 50px;
    padding-right: 50px;
  }

  &.child {
    margin-left: 90px;
    margin-right: 50px;
  }

  ${media.smallerThanMinTablet`
    &.parent {
      padding-left: 20px;
      padding-right: 20px;
    }

    &.child {
      margin-left: 40px;
      margin-right: 20px;
    }
  `}

  ${media.phone`
    &.parent {
      padding-left: 15px;
      padding-right: 15px;
    }

    &.child {
      margin-left: 15px;
      margin-right: 15px;
    }
  `}
`;

const CommmentHeaderWrapper = styled.div`
  ${media.biggerThanMinTablet`
    &.child {
      display: none;
    }
  `}
`;

const Content = styled.div`
  display: flex;
`;

const AvatarWrapper = styled.div`
  margin-right: 7px;

  ${media.smallerThanMinTablet`
    display: none;
  `}
`;

const BodyAndFooter = styled.div`
  flex: 1;
`;

const DeletedComment = styled.div`
  color: ${colors.label};
  display: flex;
  align-items: center;
  font-size: ${fontSizes.small}px;
  font-weight: 400;
  font-style: italic;
`;

const DeletedIcon = styled(Icon)`
  width: 18px;
  height: 18px;
  margin-right: 12px;
  fill: ${colors.label};
`;

interface InputProps {
  ideaId: string;
  projectId: string;
  commentId: string;
  commentType: 'parent' | 'child';
  hasBottomBorder?: boolean;
  hasChildComments?: boolean;
  last?: boolean;
  className?: string;
}

interface DataProps {
  comment: GetCommentChildProps;
  author: GetUserChildProps;
}

interface Props extends InputProps, DataProps {}

interface State {
  editing: boolean;
  translateButtonClicked: boolean;
}

class Comment extends PureComponent<Props, State> {
  static defaultProps = {
    hasBottomBorder: true,
    hasChildComment: false,
    last: false
  };

  constructor(props) {
    super(props);
    this.state = {
      editing: false,
      translateButtonClicked: false,
    };
  }

  onEditing = () => {
    this.setState({ editing: true });
  }

  onCancelEditing = () => {
    this.setState({ editing: false });
  }

  onCommentSaved = () => {
    this.setState({ editing: false });
  }

  translateComment = () => {
    const { translateButtonClicked } = this.state;

    // tracking
    if (translateButtonClicked) {
      trackEventByName(tracks.clickGoBackToOriginalCommentButton);
    } else {
      trackEventByName(tracks.clickTranslateCommentButton);
    }

    this.setState(({ translateButtonClicked }) => ({ translateButtonClicked: !translateButtonClicked }));
  }

  goToUserProfile = () => {
    const { author } = this.props;

    if (!isNilOrError(author)) {
      clHistory.push(`/profile/${author.attributes.slug}`);
    }
  }

  render() {
    const { comment, author, ideaId, projectId, commentType, hasBottomBorder, hasChildComments, last, className } = this.props;
    const { translateButtonClicked, editing } = this.state;

    if (!isNilOrError(comment)) {
      const commentId = comment.id;
      const authorId = (!isNilOrError(author) ? author.id : null);
      const lastComment = ((commentType === 'parent' && !hasChildComments) || (commentType === 'child' && last === true));
      const moderator = !isNilOrError(author) && canModerate(projectId, { data: author });

      return (
        <Container className={`${className} ${commentType} e2e-comment`}>
          <ContainerInner className={`${commentType} ${lastComment ? 'lastComment' : ''} ${hasBottomBorder ? 'hasBottomBorder' : ''}`}>
            {comment.attributes.publication_status === 'published' &&
              <>
                <CommmentHeaderWrapper className={commentType}>
                  <CommentHeader
                    projectId={projectId}
                    authorId={authorId}
                    commentId={commentId}
                    commentType={commentType}
                    commentCreatedAt={comment.attributes.created_at}
                    moderator={moderator}
                  />
                </CommmentHeaderWrapper>

                <Content>
                  {commentType === 'child' &&
                    <AvatarWrapper>
                      <Avatar
                        userId={authorId}
                        size="32px"
                        moderator={moderator}
                        onClick={this.goToUserProfile}
                        badgeBgColor={commentType === 'child' ? '#fbfbfb' : '#fff'}
                      />
                    </AvatarWrapper>
                  }

                  <BodyAndFooter>
                    <CommentBody
                      commentId={commentId}
                      commentType={commentType}
                      editing={editing}
                      moderator={moderator}
                      onCommentSaved={this.onCommentSaved}
                      onCancelEditing={this.onCancelEditing}
                      translateButtonClicked={translateButtonClicked}
                    />
                    <CommentFooter
                      className={commentType}
                      projectId={projectId}
                      ideaId={ideaId}
                      commentId={commentId}
                      commentType={commentType}
                      onEditing={this.onEditing}
                    />
                  </BodyAndFooter>
                </Content>
              </>
            }

            {comment.attributes.publication_status === 'deleted' &&
              <DeletedComment>
                <DeletedIcon name="delete" />
                <FormattedMessage {...messages.commentDeletedPlaceholder} />
              </DeletedComment>
            }
          </ContainerInner>
        </Container>
      );
    }

    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  comment: ({ commentId, render }) => <GetComment id={commentId}>{render}</GetComment>,
  author: ({ comment, render }) => <GetUser id={get(comment, 'relationships.author.data.id')}>{render}</GetUser>
});

export default (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <Comment {...inputProps} {...dataProps} />}
  </Data>
);
