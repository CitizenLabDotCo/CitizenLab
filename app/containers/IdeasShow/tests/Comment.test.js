import React from 'react';
import { fromJS } from 'immutable';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import state from './state.sample.json';
import Comments from '../components/comments';

describe('<Comment />', () => {
  it('should render a list of nested comments', () => {
    const comments = renderer.create(
      <Provider store={configureMockStore([])(fromJS(state))}>
        <Comments />
      </Provider>
    ).toJSON();
    expect(comments).toMatchSnapshot();
  });
});
