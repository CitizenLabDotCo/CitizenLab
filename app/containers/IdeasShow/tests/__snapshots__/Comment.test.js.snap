// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> should render a list of nested comments 1`] = `
<div
  className={undefined}
>
  <div
    className="ui minimal comments"
  />
</div>
`;
